### DISCUSSION TITLE
\sources\$oem$\  Directory

### DISCUSSION THREAD

**Discussion by mightymo77** [ventoy/Ventoy#1208] [Ideas]
Hello Ventoy. When extracting a Windows ISO, you can add **\sources\ $oem$\ $1\Users\Public\Desktop** (no spaces) directory, and anything you add within that Desktop folder will appear on the Windows installation. Is it possible to add such a directory within Ventoy so that it functions in the same manner? Perhaps: **D:\ $oem$\ $1\Users\Public\Desktop** where D: is the Ventoy USB drive?

**Comment by ventoy** [comment]
You can't do this directly through Ventoy.
But you can do it through unattended.xml by auto install plugin.

unattended.xml like this:
```
<?xml version="1.0" encoding="utf-8"?>
<unattend xmlns="urn:schemas-microsoft-com:unattend">
    <settings pass="oobeSystem">
        <component name="Microsoft-Windows-Shell-Setup" processorArchitecture="amd64" publicKeyToken="31bf3856ad364e35" language="neutral" versionScope="nonSxS" xmlns:wcm="http://schemas.microsoft.com/WMIConfig/2002/State" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
			<FirstLogonCommands>
                <SynchronousCommand wcm:action="add">
                    <Description>CopyWhenFirstBoot</Description>
                    <CommandLine>cmd /q /c "FOR %i IN (C D E F G H I J K L N M O P Q R S T U V W X Y Z) DO IF EXIST %i:\MY_COPY_XXX.bat cmd /k %i:\MY_COPY_XXX.bat %i"</CommandLine>
                    <Order>1</Order>
                </SynchronousCommand>								
            </FirstLogonCommands>
        </component>        
    </settings>
</unattend>
```

Put `MY_COPY_XXX.bat` in the root of Ventoy USB drive.
In the bat script you can do as you want. (Copy files from USB to C:\ or desktop ....)

**Comment by mightymo77** [comment]
Hi, I'm having a little trouble figuring out paths. So on the root of my flash drive I have a $oem$\$1\Users\Public\Desktop folder, with some files in that Desktop folder. I want to copy that Desktop folder to C:\users\public\desktop during a Windows install. The xcopy below is not working.  

xcopy \$oem\$1\Users\Public\Desktop c:\users\public\desktop /E /H /C /I

I have the above code in Batch.bat which is in the root of the flash drive, as well as Ventoy's code in the previous post (with the "MY_COPY_XXX.bat" changed to "Batch.bat") in a unattended.xml on the root as well.

Any ideas?

**Reply by steve6375** [reply]
You need to specify the drive letter of the usb drive.
As the batch file is also on the usb drive, use
Xcopy %~d0\\$oem etc. 
 as the usb drive letter and root

**Comment by ventoy** [comment]
Actually, in the above xml when `MY_COPY_XXX.bat` is called, the drive letter is the 1st parameter. (`cmd /k %i:\MY_COPY_XXX.bat %i`)
So you can also use `xcopy %1:\$oem$\$1\Users\Public\Desktop c:\users\public\desktop /E /H /C /I`

**Comment by mightymo77** [comment]
Thank you for the pointers. I cannot get it to work though. This is what I have:

unattended.xml (on root of USB):
`<?xml version="1.0" encoding="utf-8"?>
<unattend xmlns="urn:schemas-microsoft-com:unattend">
    <settings pass="oobeSystem">
        <component name="Microsoft-Windows-Shell-Setup" processorArchitecture="amd64" publicKeyToken="31bf3856ad364e35" language="neutral" versionScope="nonSxS" xmlns:wcm="http://schemas.microsoft.com/WMIConfig/2002/State" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
			<FirstLogonCommands>
                <SynchronousCommand wcm:action="add">
                    <Description>CopyWhenFirstBoot</Description>
                    <CommandLine>cmd /q /c "FOR %i IN (C D E F G H I J K L N M O P Q R S T U V W X Y Z) DO IF EXIST %i:\Batch.bat cmd /k %i:\Batch.bat %i"</CommandLine>
                    <Order>1</Order>
                </SynchronousCommand>								
            </FirstLogonCommands>
        </component>        
    </settings>
</unattend>`

Batch.bat (on root of USB):
`xcopy %i:\$oem$\$1\Users\Public\Desktop C:\Users\Public\Desktop /E /H /C /I`

Have also tried:
`xcopy %~d0\$oem$\$1\Users\Public\Desktop C:\Users\Public\Desktop /E /H /C /I`

I do have the $oem$.. folder on root of USB as well. Do you see any errors in the above code?

**Reply by steve6375** [reply]
yes
First your XML is wrong - the  pass=OobeSystem is wrong, it needs to be in the WindowsPE pass.
See https://www.easy2boot.com/add-payload-files/windows-install-isos/unattended-installs/ for an example.
Secondly, xcopy %i:\$oem$\$1\Users\Public\Desktop  is wrong use
xcopy %1:\\$oem$\\$1\Users\Public\Desktop  as LongPanda told you in above post  (number 1 not letter i).

**Comment by mightymo77** [comment]
Yes I did not mean to type an i, it was a 1. I did change the OobeSystem to WindowsPE. Still no go. Tried a few other things such as shorten the batch file to:
'xcopy %1:\$oem$\ C:\Users\Public\Desktop\Applications /E /H /C /I'

Oh well, thank you for the help though!

**Reply by steve6375** [reply]
Unless you post exact XML and exact batch file and exact contents of usb drive it is difficult to guess what you have tested. Is you windows iso 64 bit?

**Comment by mightymo77** [comment]
I have attached my .bat and .xml file. I have tested this on a Windows 10 19044.1288 x64, as well as Windows 11 22000.318 x64 installation.

**Reply by steve6375** [reply]
I tried LongPandas XML and it worked
![image](https://user-images.githubusercontent.com/1469350/142262951-f619cede-666b-463d-b270-fb8e9b923245.png)

I attach working XML and batch file. The XML has two sections, The first runs the batch file BEFORE the target drive is formatted, so it is not really useful in this case. The second part of the XML (LongPandas example) runs the same batch file at the OOBE stage (as shown in screenshot).

So you edit the ventoy.json file to use the XML file with your Win10/11 ISO.
Boot to Ventoy
Select ISO
Select XML file
at start and near end of Setup you should see the batch file run.

I have a \$oem$\ folder on my USB drive to test with
![image](https://user-images.githubusercontent.com/1469350/142264193-10f1cfcd-0ce7-4a61-90aa-613afb4948e0.png)

You can delete the first WindowsPE section of the XML file if you don't need it (which I don't think you do in this case).


[R.zip](https://github.com/ventoy/Ventoy/files/7557222/R.zip)
